### ISSUE TITLE
Make the amount of time a channel account is locked for a configurable env variable instead of a hard coded value

### ISSUE THREAD

**Issue by gouthamp-stellar** [stellar/wallet-backend#88]

**Comment by JakeUrban** [comment]
What does it mean for the channel account to be locked? 

If the channel account is specified as the source of a transaction that hasn't been accepted by the network yet, and the same channel account is "unlocked" and then used as the source for a different transaction that gets accepted by the network prior to the original, the original transaction is invalidated.

That functionality doesn't seem correct. I would think that channel accounts are locked until the transaction they're assigned to expires. If that were the case, there would be no hard-coded or configurable value for how long channel accounts are locked.

**Comment by gouthamp-stellar** [comment]
Locking/assigning a channel account to a transaction and then un-locking it when the transaction is successfully submitted to the network is definitely one way of doing this. This will involve adding an additional column to the channel_accounts table for the transaction it is currently assigned to. We will also need to make sure that the channel account gets unlocked when the transaction is actually submitted to the network

Another option I was thinking about was to have a channel account locked for the entire duration (timebounds) of the transaction. That will effectively lock a channel account to a transaction without explicitly assigning it to one. The advantage here is that we won't need that additional column. You can still explicitly unlock the channel account if/when the transaction is resolved before it times out (before it's timebounds), but you don't have to as it will naturally unlock itself given enough time

**Comment by aristidesstaffieri** [comment]
> Locking/assigning a channel account to a transaction and then un-locking it when the transaction is successfully submitted to the network is definitely one way of doing this. This will involve adding an additional column to the channel_accounts table for the transaction it is currently assigned to. We will also need to make sure that the channel account gets explicitly unlocked when the transaction is actually submitted to the network
> 
> Another option I was thinking about was to have a channel account locked for the entire duration (timebounds) of the transaction. That will effectively lock a channel account to a transaction without explicitly assigning it to one. The advantage here is that we won't need that additional column. You can still explicitly unlock the channel account if/when the transaction is resolved before it times out (before it's timebounds), but you don't have to as it will naturally unlock itself given enough time

Yeah for sure that does sound like a change that touches less surface area. One downside I see to this is that the timebounds are a maximum boundary for the transaction but they can and often will resolve in less time, so we essentially bottleneck the account by the max time it could take and not the actual time it is taking. 

Do you think that matters in this case?

**Comment by gouthamp-stellar** [comment]
Yeap, that's true. If we go with timebound based locking, it is possible that we will be locking a channel account for longer than it needs to be locked. To get over that though, I think we'd just create more channel accounts than needed when wallet backend starts up

If you do decide to go with the approach where channel accounts are locked to transactions, you can just unlock the channel accounts in the webhook channel after the final result for the transaction has been delivered to the client via the webhook

The thing is we need to come up with an approach here that works for both TSS and the account creation sponsorship endpoint, which also uses channel accounts. In the case of account creation sponsorship, we create a transaction with the channel account as the source and send that back to the client for them to submit to the network on their own. If you lock a channel account to this transaction, it will stay locked to that transaction forever unless the client uses TSS to submit their transaction

Maybe a hybrid approach is best - one where a channel account is locked to a transaction, but if it has been locked to transaction for a long time, then we consider it unlocked

**Comment by JakeUrban** [comment]
> If you lock a channel account to this transaction, it will stay locked to that transaction forever unless the client uses TSS to submit their transaction

Can we unlock the channel account based on the transaction's timebounds by default but opportunistically unlock channel accounts when they're submitted by the TSS? I don't think we should continue doing what we're doing where we make an assumption that its safe to unlock a channel account because some amount of time has passed.

**Comment by gouthamp-stellar** [comment]
We can lock a channel account for the entire duration of a transaction's timebounds, and then release the channel account once we get either a PENDING status or a response error code which is final (like malformed tx) when calling RPC's sendTransaction. This way, we won't be locking channel accounts for long periods of time when their corresponding transactions are successfully submitted to the network via TSS before their timebounds end, and if for some reason a transaction never gets to be resolved by TSS, its channel account will be released once the transaction is out of its timebounds

**Comment by JakeUrban** [comment]
We can't unlock transactions if they're `PENDING`, that means the transaction is in the queue but not yet included in a ledger. If we were to submit another transaction with the same channel account it would be rejected by core because it would have the same sequence number of a transaction already in the queue.

We need to wait until we get a final status (success or failure) either when we initially submit the tx (in the malformed example) or when we ingest the result.

**Comment by gouthamp-stellar** [comment]
Ahh, good point! Yes, in that case, we can unlock the channel account in either the ingestor or in the webhook channel when we send a transaction's result back to the client via a webhook

**Comment by aristidesstaffieri** [comment]
> Ahh, good point! Yes, in that case, we can unlock the channel account in either the ingestor or in the webhook channel when we send a transaction's result back to the client via a webhook

If we did it in ingestor would we have another ingest step that's like `releaseAccounts` and that gets all rows from the transaction table that are `status == success || fail` and release those locks?

**Comment by gouthamp-stellar** [comment]
If you were to do it in ingest.go, I'd recommend that you do it in the processTSSTransactions() function. You'd release the channel account locked to this transaction: https://github.com/stellar/wallet-backend/blob/main/internal/services/ingest.go#L240

However, it might be better if you actually do it in the webhook channel, somewhere in the receive function: https://github.com/stellar/wallet-backend/blob/main/internal/tss/channels/webhook_channel.go#L57

This will cover both cases either when we initially submit the tx and get a final error response (like malformedtx) when submitting a transaction or when we ingest the result. Both of those use the webhook channel

**Comment by aristidesstaffieri** [comment]
cool thanks for the discussion, I'm going to try to set the lock to the timebounds and unlock early in the webhook receive channel.

**Comment by aristidesstaffieri** [comment]
@gouthamp-stellar and I chatted through it and here's the recommendation - 

Channel account signature clients will now take the maxTime from the tx timebounds and lock the account until that time.
We will add a new column `tx_hash` to the accounts table and set this to the tx hash that the timebounds came from at the time we lock.
In the webhook's receive channel we will look up accounts by tx hash and unlock accounts who are locked to transactions that have finalized.
